lib/graphql/query/serial_execution/selection_resolution.rb

```diff
@@ -87,26 +66,29 @@ def fragment_type_can_apply?(ast_fragment)
 
         def merge_fields(field1, field2)
           field_type = query.schema.get_field(type, field2.name).type.unwrap
-
-          if field_type.kind.fields?
-            # create a new ast field node merging selections from each field.
-            # Because of static validation, we can assume that name, alias,
-            # arguments, and directives are exactly the same for fields 1 and 2.
-            GraphQL::Language::Nodes::Field.new(
-              name: field2.name,
-              alias: field2.alias,
-              arguments: field2.arguments,
-              directives: field2.directives,
-              selections: field1.selections + field2.selections
-            )
-          else
-            field2
-          end
+          return field2 unless field_type.kind.fields?
```

Comment:
Style-wise, I shy away from early return when it separates the structure of code from the meaning. Like, the meaning of this method is, "if there are fields to merge, we should merge them." So prefer code that matches that phrasing. 

(Of course, maybe someone else thinks of this method as, "unless there are fields to merge, return one of the fields".)

But they're both valid and IMO, as long as it's well-tested & split into meaningful units of code, I'd rather not delay progress to bicker about style!